rust/src/utils.rs

```diff
@@ -268,6 +268,94 @@ pub(crate) fn openat_to_dirfd(f: &openat::Dir) -> Result<cap_std::fs::Dir> {
     Ok(r)
 }
 
+/// Appends `part` to `base` in a way such that only characters that
+/// can be used in a D-Bus object path will be used. E.g. a character
+/// not in `[A-Z][a-z][0-9]_` will be escaped as `_HEX` where HEX is a
+/// two-digit hexadecimal number.
+///
+/// Note that his mapping is not bijective - e.g. you cannot go back
+/// to the original string.
+///
+/// # Arguments
+///
+/// * `base` - The base object path (without trailing '/').
+/// * `segment` - The UTF-8 string.
+pub(crate) fn append_to_object_path(base: &mut String, segment: impl AsRef<str>) {
+    for c in segment.as_ref().as_bytes() {
+        match *c {
+            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_' => base.push(*c as char),
+            b'/' | b'-' => base.push('_'),
+            c => base.push_str(format!("_{c:02X}").as_str()),
+        }
+    }
+}
+
+/// Determines if a given string is a valid D-Bus object path.
+///
+/// A valid object path starts with / followed by zero or more sequences
+/// of characters separated by / characters. Each sequence must contain
+/// only the characters [A-Z][a-z][0-9]_. No sequence (including the one
+/// following the final / character) may be empty.
+pub(crate) fn is_object_path(path: &impl AsRef<str>) -> bool {
```

Comment:
At first I was confused about this logic because I didn't see it in the original C code.
Afterwards I realized this is a fresh implementation of `g_variant_is_object_path()`, which we were using in an assert.
I think we can directly keep using <https://docs.rs/glib-sys/latest/glib_sys/fn.g_variant_is_object_path.html> here.
It doesn't seem to have a safe wrapper at the moment, but we can wrap it manually on our side for now.
I've opened https://github.com/gtk-rs/gtk-rs-core/issues/622 to add the wrapper directly in the `glib` crate.